system/src/asset_manager.cpp

```diff
@@ -0,0 +1,487 @@
+    reader.validate();
+    CHECK_TRUE(reader.isValid(), SYSTEM_ERROR_BAD_DATA);
+
+    auto info = reader.asset();
+    LOG(INFO, "Storing asset %s (hash=%s) size=%u original size=%u", info.name().c_str(), info.hash().toString().c_str(), reader.size(), reader.originalSize());
+
+    CHECK(clearUnusedAssets());
+
+    CHECK(stream.seek(0));
+    char tmp[256];
+
+    const auto fs = filesystem_get_instance(FILESYSTEM_INSTANCE_ASSET_STORAGE, nullptr);
+    CHECK_TRUE(fs, SYSTEM_ERROR_FILE);
+    fs::FsLock lock(fs);
+
+    lfs_file_t file = {};
+    lfs_remove(&fs->instance, info.name().c_str());
+    CHECK_TRUE(lfs_file_open(&fs->instance, &file, info.name().c_str(), LFS_O_WRONLY | LFS_O_CREAT | LFS_O_APPEND) == 0, SYSTEM_ERROR_FILE);
+    SCOPE_GUARD({     
+        lfs_file_close(&fs->instance, &file);
+    });
+
+    while (stream.availForRead() > 0) {
+        int read = stream.read(tmp, sizeof(tmp));
+        if (read < 0 && read != SYSTEM_ERROR_END_OF_STREAM) {
+            return read;
+        }
+        if (read == SYSTEM_ERROR_END_OF_STREAM) {
+            break;
+        }
+        CHECK(lfs_file_write(&fs->instance, &file, tmp, (size_t)read));
+    }
+
+    CHECK(setConsumerState(ASSET_MANAGER_CONSUMER_STATE_WANT));
+    return 0;
+}
+
+int AssetManager::setNotifyHook(asset_manager_notify_hook hook, void* context) {
+    hook_ = hook;
+    hookContext_ = context;
+    return 0;
+}
+
+int AssetManager::notifyIfNeeded() {
+    if (!hook_) {
+        return 0;
+    }
+    asset_manager_consumer_state state = ASSET_MANAGER_CONSUMER_STATE_NONE;
+    auto& cache = services::SystemCache::instance();
+    CHECK(cache.get(services::SystemCacheKey::ASSET_MANAGER_CONSUMER_STATE, &state, sizeof(state)));
+    if (state == ASSET_MANAGER_CONSUMER_STATE_WANT) {
+        hook_(hookContext_);
+    }
+
+    return 0;
+}
+
+int AssetManager::setConsumerState(asset_manager_consumer_state state) {
+    return services::SystemCache::instance().set(services::SystemCacheKey::ASSET_MANAGER_CONSUMER_STATE, &state, sizeof(state));
+}
+
+int AssetManager::prepareForOta(size_t size, unsigned flags, int moduleFunction) {
+    const size_t blockSize = EXTERNAL_FLASH_ASSET_STORAGE_SIZE / EXTERNAL_FLASH_ASSET_STORAGE_PAGE_COUNT;
+    const size_t freeBlocksSize = FREE_BLOCKS_REQUIRED * blockSize;
+    if (moduleFunction == MODULE_FUNCTION_ASSET && size > (EXTERNAL_FLASH_ASSET_STORAGE_SIZE - freeBlocksSize)) {
+        return SYSTEM_ERROR_TOO_LARGE;
+    }
+#if HAL_PLATFORM_ASSETS_OTA_OVERLAP
+    if (size > EXTERNAL_FLASH_OTA_SAFE_LENGTH) {
+        LOG(WARN, "Receiving module of size %u (function = %d) will invalidate asset storage", size, moduleFunction);
+        // We cannot receive an asset that will also cause the asset filesystem to be overwritten
+        if (moduleFunction == MODULE_FUNCTION_ASSET) {
+            return SYSTEM_ERROR_TOO_LARGE;
+        }
+        // Otherwise unmount/invalidate filesystem
+        CHECK(formatStorage(false /* remount */));
```

Comment:
Note that `prepareForOta` is called after the update manager's `startUpdate` which erases the OTA section. It likely wasn't intended to work that way.

Also, what would happen if the app tries to read an asset after the asset filesystem was invalidated? Would the app get an error or cause a UB? Is that a supported use case at all?